lib/iris/coords.py

```diff
@@ -1948,24 +1948,35 @@ def collapsed(self, dims_to_collapse=None):
         if np.issubdtype(self.dtype, np.str_):
             # Collapse the coordinate by serializing the points and
             # bounds as strings.
-            def serialize(x):
-                return "|".join([str(i) for i in x.flatten()])
+            def serialize(x, axis):
+                if axis is None:
+                    return "|".join(str(i) for i in x.flatten())
+                # np.apply_along_axis does not work with str.join, so we
+                # need to loop through the array directly. First move (possibly
+                # multiple) axis of interest to trailing dim(s), then make a 2D
+                # array we can loop through.
+                work_array = np.moveaxis(x, axis, range(-len(axis), 0))
+                out_shape = work_array.shape[: -len(axis)]
+                work_array = work_array.reshape(np.prod(out_shape), -1)
+
+                joined = []
+                for arr_slice in work_array:
+                    joined.append(serialize(arr_slice, None))
+
+                return np.array(joined).reshape(out_shape)
 
             bounds = None
             if self.has_bounds():
-                shape = self._bounds_dm.shape[1:]
                 bounds = []
-                for index in np.ndindex(shape):
-                    index_slice = (slice(None),) + tuple(index)
-                    bounds.append(serialize(self.bounds[index_slice]))
-                dtype = np.dtype("U{}".format(max(map(len, bounds))))
-                bounds = np.array(bounds, dtype=dtype).reshape((1,) + shape)
-            points = serialize(self.points)
-            dtype = np.dtype("U{}".format(len(points)))
+                for index in range(self.nbounds):
+                    bounds_slice = np.take(self.bounds, index, axis=-1)
+                    serialized = serialize(bounds_slice, dims_to_collapse)
+                    bounds.append(serialized)
+                # Make sure bounds dim comes last.
+                bounds = np.moveaxis(bounds, 0, -1)
+            points = serialize(self.points, dims_to_collapse)
             # Create the new collapsed coordinate.
-            coord = self.copy(
-                points=np.array(points, dtype=dtype), bounds=bounds
```

Comment:
I don't know why it was necessary to set the `dtype`, but the existing tests pass fine without this now.  The old code is 8 years old, so I guess numpy has moved on and is likely doing more for us automatically.